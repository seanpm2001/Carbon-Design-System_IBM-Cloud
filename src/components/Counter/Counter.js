import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { useTranslation } from "react-i18next";
import Skeleton from "./skeleton";

// Uses inclusive bounds.
function keepInBounds(value, lowerBound, upperBound) {
  if (value < lowerBound) {
    return lowerBound
  }
  if (value > upperBound) {
    return upperBound
  }
  return value
}

const Counter = ({
    /* Declare any props that this pattern can use */
    totalSteps,
    currentStep: initialStep,
  }) => {
  const [currentStep, setCurrentStep] = useState(keepInBounds(initialStep, 0, totalSteps));
  const [previousStep, setPreviousStep] = useState(currentStep);

  useEffect(() => {
    let newStep = keepInBounds(initialStep, 0, totalSteps);
    if (newStep !== currentStep) {
      setPreviousStep(currentStep);
      setCurrentStep(newStep);
    }
  }, [initialStep, currentStep, totalSteps]);

  const { t } = useTranslation("Counter");

  return (
      <div className="pal--counter" role="status">
      <div className="pal--counter__numbers">
        {/* {currentStep !== previousStep
          ? <h1 key={previousStep} className="pal--counter__previous-step pal--counter__number-exit">{previousStep}</h1>
          : null
        } */}
        {currentStep > previousStep
          ? <h1 key={previousStep} className="pal--counter__previous-step pal--counter__increase_number-exit">{previousStep}</h1>
          : null
        }
        {currentStep < previousStep
          ? <h1 key={previousStep} className="pal--counter__previous-step pal--counter__decrease_number-exit">{previousStep}</h1>
          : null
        }
        <h1 key={currentStep} className="pal--counter__current-step pal--counter__increase_number-enter">{currentStep}</h1>
        <h1 className="pal--counter__total-steps">/ {totalSteps}</h1>
      </div>
      <p className="pal--counter__label" aria-label={t("steps")}>{t("steps")}</p>
      <hr className="pal--counter__divider"></hr>
      </div>
  );
};

// components should export a skeleton
Counter.Skeleton = Skeleton;

Counter.propTypes = {
  /**
   * The total number of steps in the process.
   */
  totalSteps: PropTypes.number.isRequired, // 
  /**
   * The number of steps the user is currently on, or the starting number if not 0. Use this to increase the current step number.
   */
  currentStep: PropTypes.number, // 
};

Counter.defaultProps = {
  /* Default value for any non required props for this pattern */
  totalSteps: 5,
  currentStep: 0,
};

export default Counter;