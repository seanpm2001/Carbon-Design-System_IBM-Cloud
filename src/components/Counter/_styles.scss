@use "@carbon/react/scss/type";
@use '@carbon/react/scss/reset';
@import "@carbon/react/scss/colors";
@import "@carbon/react/scss/motion";
@import "@carbon/react/scss/spacing";
@import "@carbon/react/scss/themes";
@import "@carbon/react/scss/theme";
@import "@carbon/react/scss/config";

// @include component('pal--counter') {
.pal--counter {
    text-align: left;
    margin: 16px;
    transition: width 100ms;

    &__numbers {
        display: flex;
        margin-bottom: $spacing-05;
        align-items: baseline;
    }

    &__previous-step {
        @include type.type-style('productive-heading-07');
        // font-family: 'IBM Plex Sans';
        // font-style: normal;
        font-weight: 300;
        // font-size: 54px;
        // line-height: 64px;
        color: $text-primary;
        position: relative;
        width: 0;
        height: 0;
    }

    &__current-step {
        @include type.type-style('productive-heading-07');
        // font-family: 'IBM Plex Sans';
        // font-style: normal;
        font-weight: 300;
        // font-size: 54px;
        // line-height: 64px;
        color: $text-primary;
    }

    &__number-exit {
        @media (prefers-reduced-motion: reduce) {
            animation-duration: 600ms;
        }

        animation: 400ms motion(exit, expressive) 0ms 1 forwards numberExit;

        @keyframes numberExit {
            0% {
                transform: translateY(0);
                opacity: 1;
            }

            60% {
                opacity: 0;
            }

            100% {
                transform: translateY(-4.5rem);
                opacity: 0;
                display: none;
                visibility: hidden;
            }

        }
    }

    &__number-enter {
        @media (prefers-reduced-motion: reduce) {
            animation-duration: 600ms;
        }

        animation: 450ms motion(entrance, expressive) 0ms 1 forwards numberEnter;

        @keyframes numberEnter {
            0% {
                transform: translateY(4.5rem);
                opacity: 0;
            }

            20% {
                opacity: 0;
            }

            100% {
                transform: translateY(0);
                opacity: 1;
            }

        }
    }

    &__total-steps {
        @include type.type-style('heading-04');
        color: $text-secondary;
        margin-left: 6px;
    }

    &__label {
        @include type.type-style('label-02');
        color: $text-primary;
    }

    &__divider {
        border: 0;
        margin: 14px 0 20px 0;
        text-align: left;
        width: 216px;
        height: 1px;
        background: $border-subtle-01;
    }
}

// }