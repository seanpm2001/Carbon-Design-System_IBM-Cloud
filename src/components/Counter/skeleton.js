import React from "react";

const Skeleton = () => {
  return <div className="pal--counter__skeleton>">
          <h1 className="pal--counter__total-steps">0 / 5</h1>
          <p className="pal--counter__label">Steps complete</p>
          <hr className="pal--counter__divider"></hr>
      </div>;
};

export default Skeleton