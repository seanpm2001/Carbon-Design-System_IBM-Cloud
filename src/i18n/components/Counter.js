import en from "../../locales/en/Counter.json";
// import de from "../../locales/de/Counter.json";
// import es from "../../locales/es/Counter.json";
// import fr from "../../locales/fr/Counter.json";
// import it from "../../locales/it/Counter.json";
// import ja from "../../locales/ja/Counter.json";
// import ko from "../../locales/ko/Counter.json";
// import pt_br from "../../locales/pt_br/Counter.json";
// import zh_cn from "../../locales/zh_cn/Counter.json";
// import zh_tw from "../../locales/zh_tw/Counter.json";

export default {
  en,
  // de,
  // es,
  // fr,
  // it,
  // ja,
  // ko,
  // pt_br,
  // zh_cn,
  // zh_tw,
};
