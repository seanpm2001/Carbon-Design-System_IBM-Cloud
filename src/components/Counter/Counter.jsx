import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import "./Counter.scss";

// Uses inclusive bounds.
function keepInBounds(value, lowerBound, upperBound) {
  if (value < lowerBound) {
    return lowerBound
  }
  if (value > upperBound) {
    return upperBound
  }
  return value
}

const Counter = ({
    /* Declare any props that this pattern can use */
    totalSteps,
    currentStep: initialStep,
  }) => {
  const [currentStep, setCurrentStep] = useState(keepInBounds(initialStep, 0, totalSteps));
  const [previousStep, setPreviousStep] = useState(currentStep);

  useEffect(() => {
    let newStep = keepInBounds(initialStep, 0, totalSteps);
    if (newStep !== currentStep) {
      setPreviousStep(currentStep);
      setCurrentStep(newStep);
    }
  }, [initialStep]);

  return (
      <div className="pal--counter" role="status">
      <div className="pal--counter__numbers">
        {currentStep !== previousStep
          ? <h1 key={previousStep} className="pal--counter__previous-step pal--counter__number-exit">{previousStep}</h1>
          : null
        }
        <h1 key={currentStep} className="pal--counter__current-step pal--counter__number-enter">{currentStep}</h1>
        <h1 className="pal--counter__total-steps">/ {totalSteps}</h1>
      </div>
      <p className="pal--counter__label" aria-labelledby>Steps complete</p>
      <hr className="pal--counter__divider"></hr>
      </div>
  );
};

// components should export a skeleton
Counter.Skeleton = () => {
  /* Build skeleton component structure here */
  // TODO Update Below!
  return <div className="pal--counter__skeleton>">
          <h1 className="pal--counter__total-steps">0 / 5</h1>
          <p className="pal--counter__label">Steps complete</p>
          <hr className="pal--counter__divider"></hr>
      </div>;
};

Counter.propTypes = {
  totalSteps: PropTypes.number.isRequired, // Total steps in the flow
  currentStep: PropTypes.number, // Optional prop for changing starting step number
};

Counter.defaultProps = {
  /* Default value for any non required props for this pattern */
  totalSteps: 5,
  currentStep: 0,
};

export default Counter;