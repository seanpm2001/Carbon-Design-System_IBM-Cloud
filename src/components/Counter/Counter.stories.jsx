import React, { useState } from 'react';
import { Button } from "@carbon/react";
import Counter from './Counter';

export default {
  component: Counter,
  title: 'Components/Counter'
};

//👇 We create a “template” of how args map to rendering
const Template = args => {
  const {currentStep, ...otherArgs} = args;
  const [step, setStep] = useState(currentStep);
  function addSteps(delta) {
    let newStep = step + delta;
    // Limit our step state to what can be displayed,
    // otherwise button clicks may take more increments/decrements than is visible.
    if (newStep < 0) {
      newStep = 0;
    } else if (newStep > args.totalSteps) {
      newStep = args.totalSteps;
    }
    setStep(newStep);
  }
  return <>
    <Counter currentStep={step} {...otherArgs} />
    <Button kind='tertiary' size='sm' onClick={() => addSteps(1)}>
        Increase
    </Button>
    <Button kind='tertiary' size='sm'onClick={() => addSteps(-1)}>
        Decrease
    </Button>
  </>
}

//👇 Each story then reuses that template
export const Hovers = Template.bind({});

Hovers.args = {totalSteps:5, currentStep:0};
