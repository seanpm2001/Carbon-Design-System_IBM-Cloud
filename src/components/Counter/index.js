import Counter from './Counter';
import CounterSkeleton from './skeleton';

export { Counter, CounterSkeleton };

export default Counter;