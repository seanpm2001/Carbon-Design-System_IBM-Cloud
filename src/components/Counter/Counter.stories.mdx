import { Meta, Canvas, Story, ArgsTable } from "@storybook/addon-docs";
import { useState } from 'react';
import { Button } from "@carbon/react";
import Counter from './Counter';
import "./_styles.scss";

<Meta
  title="Experimental/Counter"
  component={Counter}
  argTypes={{
    totalSteps: { type: { name: 'number' } }, currentStep: { type: { name: 'number' } },
  }}
  parameters={{ status: { type: "experimental" } }}
/>

# Counter

export const Template = args => {
  const {currentStep, ...otherArgs} = args;
  const [step, setStep] = useState(currentStep);
  function addSteps(delta) {
    let newStep = step + delta;
    // Limit our step state to what can be displayed,
    // otherwise button clicks may take more increments/decrements than is visible.
    if (newStep < 0) {
      newStep = 0;
    } else if (newStep > args.totalSteps) {
      newStep = args.totalSteps;
    }
    setStep(newStep);
  }
  return <>
    <Counter currentStep={step} {...otherArgs} />
    <Button kind='tertiary' size='sm' onClick={() => addSteps(1)}>
        Increase
    </Button>
    <Button kind='tertiary' size='sm'onClick={() => addSteps(-1)}>
        Decrease
    </Button>
    </>
}

### Default

<Canvas>
  <Story
    name="Default"
    args={{totalSteps:5, currentStep:0}}
  >
    {Template.bind({})}
  </Story>
</Canvas>

## Description

Counter is a component in the Cloud PAL Experimental style. The counter monitors the userâ€™s progress through a series of steps. 

It guides them toward the completion of a specific process and shows the numeric value of the completed steps in relation to how many steps there are in total.

Motion can be reduced with system or browser settings.

## Usage

1.  Install the [@carbon/carbon-for-ibm-cloud](https://pages.github.ibm.com/cdai-design/cloud-pal/cloudpal-guidance/cloudpal-v2/migration-guides/#installing-and-using-the-new-packages) library
2.  Import the styles:

    ```js
    @import '@carbon/carbon-for-ibm-cloud/Components/Counter/styles';
    ```

3.  Import the component:

    ```js
    import { Counter } from "@carbon/carbon-for-ibm-cloud/Components";
    ```

4.  Render the component.

For general component usage review the [Components Docs](https://pages.github.ibm.com/ibmcloud/pal/development-guidelines/components/overview).

## Props

<ArgsTable of={Counter} />

## Contributors

- Stephanie Cree
- Lisa Kaiser
- Asha Newsom
- Max Hammer