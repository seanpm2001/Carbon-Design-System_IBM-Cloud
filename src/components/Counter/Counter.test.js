import React from 'react';
import { render } from '../../../test-utils';

import Counter from '.';

describe('<Counter />', () => {
  it('renders the component as expected', () => {
    const { baseElement } = render(
      <Counter
        totalSteps={5}
        currentStep={3}
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });
  
  it('renders the new props as expected', () => {
    const { baseElement } = render(
      <Counter
        totalSteps={0}
        currentStep={10}
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });
  it('does not change out of bound props as expected', () => {
    const { baseElement } = render(
      <Counter
        totalSteps={11}
        currentStep={10}
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });
  it('does not change out of bound props as expected', () => {
    const { baseElement } = render(
      <Counter
        totalSteps={-1}
        currentStep={5}
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });
  it('does not apply strings as numbers', () => {
    const { baseElement } = render(
      <Counter
        totalSteps="not a number"
        currentStep="another string"
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });
});
