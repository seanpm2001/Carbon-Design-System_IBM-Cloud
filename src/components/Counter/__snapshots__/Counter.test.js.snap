// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Counter /> does not apply strings as numbers 1`] = `
<body>
  <div>
    <div
      class="cds--white cds--layer-one"
    >
      <div
        class="pal--counter"
        role="status"
      >
        <div
          class="pal--counter__numbers"
        >
          <h1
            class="pal--counter__current-step pal--counter__number-enter"
          >
            another string
          </h1>
          <h1
            class="pal--counter__total-steps"
          >
            / 
            not a number
          </h1>
        </div>
        <p
          aria-label="Steps complete"
          class="pal--counter__label"
        >
          Steps complete
        </p>
        <hr
          class="pal--counter__divider"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`<Counter /> does not change out of bound props as expected 1`] = `
<body>
  <div>
    <div
      class="cds--white cds--layer-one"
    >
      <div
        class="pal--counter"
        role="status"
      >
        <div
          class="pal--counter__numbers"
        >
          <h1
            class="pal--counter__current-step pal--counter__number-enter"
          >
            10
          </h1>
          <h1
            class="pal--counter__total-steps"
          >
            / 
            11
          </h1>
        </div>
        <p
          aria-label="Steps complete"
          class="pal--counter__label"
        >
          Steps complete
        </p>
        <hr
          class="pal--counter__divider"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`<Counter /> does not change out of bound props as expected 2`] = `
<body>
  <div>
    <div
      class="cds--white cds--layer-one"
    >
      <div
        class="pal--counter"
        role="status"
      >
        <div
          class="pal--counter__numbers"
        >
          <h1
            class="pal--counter__current-step pal--counter__number-enter"
          >
            -1
          </h1>
          <h1
            class="pal--counter__total-steps"
          >
            / 
            -1
          </h1>
        </div>
        <p
          aria-label="Steps complete"
          class="pal--counter__label"
        >
          Steps complete
        </p>
        <hr
          class="pal--counter__divider"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`<Counter /> renders the component as expected 1`] = `
<body>
  <div>
    <div
      class="cds--white cds--layer-one"
    >
      <div
        class="pal--counter"
        role="status"
      >
        <div
          class="pal--counter__numbers"
        >
          <h1
            class="pal--counter__current-step pal--counter__number-enter"
          >
            3
          </h1>
          <h1
            class="pal--counter__total-steps"
          >
            / 
            5
          </h1>
        </div>
        <p
          aria-label="Steps complete"
          class="pal--counter__label"
        >
          Steps complete
        </p>
        <hr
          class="pal--counter__divider"
        />
      </div>
    </div>
  </div>
</body>
`;

exports[`<Counter /> renders the new props as expected 1`] = `
<body>
  <div>
    <div
      class="cds--white cds--layer-one"
    >
      <div
        class="pal--counter"
        role="status"
      >
        <div
          class="pal--counter__numbers"
        >
          <h1
            class="pal--counter__current-step pal--counter__number-enter"
          >
            0
          </h1>
          <h1
            class="pal--counter__total-steps"
          >
            / 
            0
          </h1>
        </div>
        <p
          aria-label="Steps complete"
          class="pal--counter__label"
        >
          Steps complete
        </p>
        <hr
          class="pal--counter__divider"
        />
      </div>
    </div>
  </div>
</body>
`;
